CriteoPublisherSdk/Sources/Logging/CR_LogMessage.m

```diff
@@ -0,0 +1,96 @@
+//
+//  CR_LogMessage.m
+//  CriteoPublisherSdk
+//
+//  Copyright © 2018-2020 Criteo. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import "CR_LogMessage.h"
+
+@implementation CR_LogMessage
+
+- (instancetype)initWithTag:(NSString *)tag
+                   severity:(CR_LogSeverity)severity
+                       file:(NSString *)file
+                       line:(NSUInteger)line
+                   function:(NSString *)function
+                  exception:(NSException *_Nullable)exception
+                    message:(NSString *)message {
+  self = [super init];
+  if (self) {
+    _tag = tag;
+    _severity = severity;
+    _file = file;
+    _line = line;
+    _function = function;
+    _timestamp = [NSDate date];
+    _exception = exception;
+    _message = message;
+  }
+
+  return self;
+}
+
++ (instancetype)messageWithTag:(NSString *)tag
+                      severity:(CR_LogSeverity)severity
+                          file:(const char *)file
+                          line:(int)line
+                      function:(const char *)function
+                     exception:(NSException *_Nullable)exception
+                        format:(NSString *)format, ... {
+  va_list args;
+  va_start(args, format);
+  NSString *message = [[NSString alloc] initWithFormat:format arguments:args];
+  va_end(args);
+  return [[self alloc] initWithTag:tag
+                          severity:severity
+                              file:[NSString stringWithUTF8String:file]
+                              line:(NSUInteger)line
+                          function:[NSString stringWithUTF8String:function]
+                         exception:exception
+                           message:message];
+}
+
+- (NSString *)severityLabel {
+  switch (self.severity) {
+    case CR_LogSeverityDebug:
+      return @"🐛DEBUG";
+    case CR_LogSeverityError:
+      return @"🔴ERROR";
+    case CR_LogSeverityWarning:
+      return @"🟠WARN";
+    case CR_LogSeverityInfo:
+      return @"ℹ️INFO";
+    default:
+      return nil;
+  }
+}
+
+- (NSString *)description {
```

Comment:
Do you think it's worth covering this method with tests?

Reply:
I don't think so as it is only implemented for development purposes, and isn't in production code path.
I'd rather remove it than testing.

For information, `-[NSObject description]` is the Objective-C equivalent of Java's `toString()` one, which is informational for for debugging purposes, not worth to test imo most of the time, exceptions being like for the other tests, a client (TS in this case) that request specifically some data to be logged.

Reply:
Ok, my bad, thought it was returning the value that was used by the logger when printing on the console.